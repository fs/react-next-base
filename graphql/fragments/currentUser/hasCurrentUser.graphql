fragment HasCurrentUser on CurrentUser {
  id
}
