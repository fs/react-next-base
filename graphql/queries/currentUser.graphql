#import "../fragments/userInfo.graphql"

query CurrentUser {
  me {
    ...UserInfo
  }
}
