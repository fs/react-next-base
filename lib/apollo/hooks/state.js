import { useQuery } from '@apollo/client';
import CurrentUser from 'graphql/queries/currentUser.graphql';

export const useCurrentUser = (prefetch = true) => {
  const { loading, data, error } = useQuery(CurrentUser, {
    fetchPolicy: prefetch ? 'cache-first' : 'cache-only',
  });

  const user = data?.me;

  return { user, loading, error };
};
