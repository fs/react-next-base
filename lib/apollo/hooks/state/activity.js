/* eslint-disable import/prefer-default-export */
import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';
import activityPageSizes from 'config/activityPageSizes';

const getFormattedActivity = (data) => {
  return data.activities.edges.map(
    ({
      node: {
        id,
        title,
        body,
        createdAt,
        event,
        user: { firstName, lastName, email, avatarUrl },
      },
    }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find((e) => e.value === event).color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = ({ beforeCursor, afterCursor, filterValue, pageSize = activityPageSizes[0] }) => {
  const events = filterValue ? [filterValue] : activityEvents.map(({ value }) => value);
  const first = afterCursor || (!afterCursor && !beforeCursor) ? pageSize : undefined;
  const last = beforeCursor ? pageSize : undefined;

  const { loading, error, data } = useQuery(Activities, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    variables: {
      events,
      last,
      before: beforeCursor,
      first,
      after: afterCursor,
    },
  });

  return {
    activities: data ? getFormattedActivity(data) : null,
    pageInfo: data ? data.activities.pageInfo : null,
    loading,
    error,
  };
};
