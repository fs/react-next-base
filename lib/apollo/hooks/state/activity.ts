import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';
import activityPageSizes from 'config/activityPageSizes';

import User from 'domain/User';
import { PageInfo, Activity } from 'interfaces/activityType';

type ActivityParams = {
  beforeCursor?: string;
  afterCursor?: string;
  filterValue?: string;
  pageSize: number;
};

type Edge = {
  cursor: string;
  node: {
    body: string;
    createdAt: string | Date;
    event: string;
    id: string | number;
    title: string;
    user: User;
  };
};

type ActivityQueryState = {
  activities: {
    edges: Edge[];
    pageInfo: PageInfo;
  };
};

type ActivityQueryVars = {
  events: string[];
  last?: number;
  before?: string;
  first?: number;
  after?: string;
};

const getFormattedActivity = (edges: Edge[]): Activity[] => {
  return edges.map(
    ({
      node: {
        id,
        title,
        body,
        createdAt,
        event,
        user: { firstName, lastName, email, avatarUrl },
      },
    }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find((e) => e.value === event)?.color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = ({
  beforeCursor,
  afterCursor,
  filterValue,
  pageSize = activityPageSizes[0],
}: ActivityParams) => {
  const events = filterValue ? [filterValue] : activityEvents.map(({ value }) => value);
  const first = afterCursor || (!afterCursor && !beforeCursor) ? pageSize : undefined;
  const last = beforeCursor ? pageSize : undefined;

  const { loading, error, data } = useQuery<ActivityQueryState, ActivityQueryVars>(Activities, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    variables: {
      events,
      last,
      before: beforeCursor,
      first,
      after: afterCursor,
    },
  });

  return {
    activities: data?.activities?.edges ? getFormattedActivity(data?.activities.edges) : null,
    pageInfo: data?.activities?.pageInfo || null,
    loading,
    error,
  };
};
