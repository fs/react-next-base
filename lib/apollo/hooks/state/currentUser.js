import { useQuery } from '@apollo/client';
import CurrentUser from 'graphql/queries/currentUser.graphql';

export const useCurrentUser = (prefetch = true) => {
  const { data, loading, error } = useQuery(CurrentUser, {
    fetchPolicy: prefetch ? 'cache-first' : 'cache-only',
  });

  return {
    user: data?.me,
    loading,
    error,
  };
};
