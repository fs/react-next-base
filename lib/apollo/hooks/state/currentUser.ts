import { useQuery } from '@apollo/client';
import CurrentUser from 'graphql/queries/currentUser.graphql';
import User from 'domain/User';

type UserData = {
  me: User;
};

export const useCurrentUser = (prefetch = true) => {
  const { data, loading, error } = useQuery<UserData>(CurrentUser, {
    fetchPolicy: prefetch ? 'cache-first' : 'cache-only',
  });

  return {
    user: data?.me,
    loading,
    error,
  };
};
