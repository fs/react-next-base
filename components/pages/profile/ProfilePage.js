import React from 'react';

import ErrorDecorator from 'decorators/ErrorDecorator';
import ErrorMessage from 'components/shared/atoms/ErrorMessage';
import WithAuth from 'lib/auth/withAuth';
import WithAuthSecurity from 'lib/auth/withAuthSecurity';
import { withApolloClient } from 'lib/withApolloClient';
import { useCurrentUser } from 'lib/apollo/hooks/state/currentUser';

import DefaultTemplate from 'components/shared/templates/DefaultTemplate';
import ProfileForm from 'components/shared/organisms/ProfileForm';
import { NotifierProvider } from 'contexts/NotifierContext';
import Notifier from 'components/shared/atoms/Notifier';

const Profile = () => {
  const { loading, error, user } = useCurrentUser();
  const profile = user || {};
  const errorMessage = error ? new ErrorDecorator(error).getMessages() : null;

  return (
    <NotifierProvider>
      {loading && <h3 data-testid="profile-loading">Loading...</h3>}
      {error && <ErrorMessage testId="profile-error">{errorMessage}</ErrorMessage>}
      {!loading && !error && (
        <DefaultTemplate data-testid="profile-page">
          <ProfileForm profile={profile} />
          <Notifier />
        </DefaultTemplate>
      )}
    </NotifierProvider>
  );
};

export default withApolloClient(WithAuth(WithAuthSecurity(Profile)));
