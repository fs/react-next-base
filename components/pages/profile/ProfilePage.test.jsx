import React from 'react';
import renderWithApolloClient from '__tests__/helpers/renderWithApolloClient';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import renderWithTheme from '__tests__/helpers/renderWithTheme';
import { useCurrentUser } from 'lib/apollo/hooks/state/currentUser';

import ProfilePage from '.';

jest.mock('components/shared/templates/DefaultTemplate', () => 'div');
jest.mock('lib/apollo/hooks/state/currentUser');
jest.mock('lib/apollo/hooks/actions/useSignIn', () => jest.fn(() => []));
jest.mock('lib/apollo/hooks/actions/useUpdateUser', () => jest.fn(() => []));
jest.mock('lib/apollo/hooks/actions/usePresignFile', () => jest.fn(() => []));

describe('Profile page', () => {
  test('should render correctly', async () => {
    // Arrange
    const mockUseCurrentUser = jest.fn(() => ({
      loading: undefined,
      error: undefined,
      user: { id: '1', email: 'user@mail.ru', firstName: 'First Name' },
    }));
    useCurrentUser.mockImplementation(mockUseCurrentUser);

    // Act
    render(renderWithTheme(renderWithApolloClient(<ProfilePage />)));
    const pageContent = screen.getByTestId('profile-page');

    // Assert
    expect(pageContent).toMatchSnapshot();
  });

  test('should show loader while loading', async () => {
    // Arrange
    const mockUseCurrentUser = jest.fn(() => ({
      loading: true,
      error: undefined,
      user: undefined,
    }));
    useCurrentUser.mockImplementation(mockUseCurrentUser);

    // Act
    render(renderWithTheme(renderWithApolloClient(<ProfilePage />)));
    const loadingContent = screen.getByTestId('profile-loading');

    // Assert
    expect(loadingContent).toMatchSnapshot();
  });

  test('should show error on error', async () => {
    // Arrange
    const error = {
      graphQLErrors: [
        {
          message: 'Current User not found or accessible',
          locations: [{ line: 11, column: 5 }],
        },
      ],
      networkError: null,
      message: 'GraphQL error: Current User not found or accessible',
    };
    const mockUseCurrentUserErrorData = jest.fn(() => ({
      loading: false,
      error,
      user: undefined,
    }));
    useCurrentUser.mockImplementation(mockUseCurrentUserErrorData);

    // Act
    render(renderWithTheme(renderWithApolloClient(<ProfilePage />)));
    const errorContent = screen.getByTestId('profile-error');

    // Assert
    expect(errorContent).toMatchSnapshot();
  });
});
