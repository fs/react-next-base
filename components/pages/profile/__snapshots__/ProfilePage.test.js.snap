// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile page should render correctly 1`] = `
.c2 {
  max-width: 40rem;
}

.c2 input,
.c2 textarea,
.c2 select,
.c2 button {
  width: 100%;
  padding: 1rem;
  font-size: 1rem;
  border-radius: 0.3rem;
  border: 1px solid rgb(179,179,179);
}

.c2 [type='file'] {
  border: none;
  padding: 0;
}

.c2 input[type='checkbox'],
.c2 input[type='radio'] {
  width: auto;
}

.c2 textarea {
  resize: vertical;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  position: relative;
  padding-top: 1.5rem;
  margin-bottom: 1rem;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
}

.c0 {
  margin-left: auto;
  margin-right: auto;
  width: 20rem;
  position: relative;
}

.c1 {
  max-width: 40rem;
  margin: 0 auto 1rem;
  line-height: 1.25;
  -webkit-letter-spacing: -0.035rem;
  -moz-letter-spacing: -0.035rem;
  -ms-letter-spacing: -0.035rem;
  letter-spacing: -0.035rem;
}

<div
  data-testid="profile-page"
>
  <div
    class="c0"
  >
    <h3
      class="c1"
      data-cy="profile-form-title"
    >
      Profile
    </h3>
    <div
      class="c2"
      data-cy="profile-update-form"
    >
      <form
        action="#"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <label
              class="c5"
              for="avatar"
            >
              Avatar
            </label>
            <input
              accept="image/*"
              data-cy="avatar"
              data-testid="avatar"
              id="avatar"
              name="avatar"
              type="file"
              value=""
            />
          </div>
          <div
            class="c4"
          >
            <label
              class="c5"
              for="firstName"
            >
              First Name
            </label>
            <input
              data-cy="first-name"
              data-testid="first-name"
              id="firstName"
              name="firstName"
              placeholder="First Name"
              type="text"
              value="First Name"
            />
          </div>
          <div
            class="c4"
          >
            <label
              class="c5"
              for="lastName"
            >
              Last Name
            </label>
            <input
              data-cy="last-name"
              data-testid="last-name"
              id="lastName"
              name="lastName"
              placeholder="Last Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="c4"
          >
            <label
              class="c5"
              for="email"
            >
              Email
            </label>
            <input
              data-cy="email"
              data-testid="email"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
              value="user@mail.ru"
            />
          </div>
          <div
            class="c4"
          >
            <label
              class="c5"
              for="password"
            >
              New Password
            </label>
            <input
              data-cy="password"
              data-testid="password"
              id="password"
              name="password"
              placeholder="New Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="c4"
          >
            <label
              class="c5"
              for="currentPassword"
            >
              Current Password
            </label>
            <input
              data-cy="current-password"
              data-testid="current-password"
              id="currentPassword"
              name="currentPassword"
              placeholder="Current Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="c4"
          >
            <input
              data-cy="update-button"
              data-testid="update-button"
              id="Update"
              name="Update"
              type="submit"
            />
          </div>
        </div>
      </form>
    </div>
  </div>
  <div
    data-cy="notifier"
  >
    <div
      class="Toastify"
    />
  </div>
</div>
`;

exports[`Profile page should show error on error 1`] = `
<div
  data-cy="profile-error"
  data-testid="profile-error"
>
  <p>
    Current User not found or accessible
  </p>
</div>
`;

exports[`Profile page should show loader while loading 1`] = `
<h3
  data-testid="profile-loading"
>
  Loading...
</h3>
`;
